File: cs/Letmecode/StringTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Text.RegularExpressions;
using System.Linq;

namespace Letmecode
{
    [TestClass]
    public class StringTest
    {

        [TestMethod]
        public void ComposeString()
        {
            Assert.AreEqual("apple orange", string.Format("{0} {1}", "apple", "orange"));
        }
        
        [TestMethod]
        public void MultilineString()
        {
            var text = @"apple
orange";
            Assert.AreEqual("apple" + System.Environment.NewLine + "orange", text);
        }

        [TestMethod]
        public void IntToString()
        {
            Assert.AreEqual("1", 1.ToString());
        }

        [TestMethod]
        public void RegexMatch()
        {
            // https://msdn.microsoft.com/en-us/library/az24scfc(v=vs.110).aspx
            Assert.AreEqual(true, new Regex("apple").IsMatch("apple orange "));
            Assert.AreEqual(true, new Regex("orange").IsMatch("apple orange "));
            Assert.AreEqual(false, new Regex("kiwi").IsMatch("apple orange "));
        }

        [TestMethod]
        public void RegexMatches()
        {
            Assert.AreEqual("apple orange",
              Regex.Match("apple orange ", "(apple).*(orange)").Groups[0].Value);
            Assert.AreEqual("apple",
              Regex.Match("apple orange ", "(apple).*(orange)").Groups[1].Value);
            Assert.AreEqual("orange",
              Regex.Match("apple orange ", "(apple).*(orange)").Groups[2].Value);
        }

        [TestMethod]
        public void RegexReplace()
        {
            Assert.AreEqual("orange apple ",
                    Regex.Replace("apple orange ", "(apple).*(orange)", "$2 $1"));
        }

        [TestMethod]
        public void Concat()
        {
            var text = "apple";
            var textBackup = text;
            text += " orange ";
            Assert.AreEqual("apple orange ", text);
            Assert.AreNotEqual(text, textBackup);
        }

        [TestMethod]
        public void ConcatMutate()
        {
            var sb = new System.Text.StringBuilder();
            sb.Append("apple");
            sb.Append(" orange ");
            Assert.AreEqual("apple orange ", sb.ToString());
        }

        [TestMethod]
        public void Length()
        {
            Assert.AreEqual(13, "apple orange ".Length);
        }

        [TestMethod]
        public void Format()
        {
            // https://msdn.microsoft.com/en-us/library/dwhawy9k(v=vs.110).aspx
            Assert.AreEqual("10.0", string.Format("{0:f1}", 10));
            // https://msdn.microsoft.com/en-us/library/8kb3ddd4(v=vs.110).aspx
            Assert.AreEqual("2000-01-01",
              new System.DateTime(2000, 1, 1).ToString("yyyy-MM-dd"));
        }

        [TestMethod]
        public void Split()
        {
            Assert.AreEqual("apple", "apple orange ".Trim().Split(null)[0]);
            Assert.AreEqual(2, "apple orange ".Trim().Split(null).Length);
            Assert.AreEqual("apple", "apple,orange ".Trim().Split(',')[0]);
        }

        [TestMethod]
        public void Strip()
        {
            Assert.AreEqual("a b", " a b ".Trim());
        }

        [TestMethod]
        public void eql()
        {
            Assert.AreEqual(true, "apple" == "apple");
        }
    }
}

File: nodejs/string_test.js
var assert = require('assert');

(function composeString() {
  assert.equal("apple orange", "apple" + " " + "orange");
})();

(function multilineString() {
  var text = [
    "apple",
    "orange",
    ""
  ].join("\n");
  assert.equal("apple\norange\n", text);
})();

(function intToString() {
  assert.equal("1", String(1));
})();

(function regexMatch() {
  // http://www.w3schools.com/js/js_regexp.asp
  assert.equal(0, "apple orange ".match(/apple/)["index"]);
  assert.equal(6, "apple orange ".match(/orange/)["index"]);
  assert.equal(null, "apple orange ".match(/kiwi/));
})();

(function regexMatches() {
  var array = ["apple orange", "apple", "orange"];
  var match = "apple orange ".match(/(apple).*(orange)/);
  assert.equal(3, match.length);
  assert.equal(array[0], match[0]);
  assert.equal(array[1], match[1]);
  assert.equal(array[2], match[2]);
})();

(function concat() {
  var text = "apple";
  var text_backup = text;
  text += " orange ";
  assert.equal("apple orange ", text);
})();

(function concatMutate() {
  // Not possible
})();

(function length() {
  assert.equal(13, "apple orange ".length);
})();

(function format() {
  assert.equal("10.0", 10.001.toFixed(1)); // round
  // http://www.w3schools.com/jsref/jsref_obj_date.asp
  var time  = new Date(2000, 1 - 1, 1, 0, 0, 0);
  var year  = time.getFullYear();
  var month = ("0" + (time.getMonth() + 1)).slice(-2);
  var date  = ("0" + time.getDate()).slice(-2);
  assert.equal("2000-01-01", [year, month, date].join("-"))
})();

(function split() {
  var array = "apple orange ".trim().split(" ");
  assert.equal(2, array.length);
  assert.equal("apple", array[0]);
  assert.equal("orange", array[1]);
})();

(function strip() {
  assert.equal("a b", " a b ".trim());
})();

(function eql() {
  assert.equal(true, "apple" == "apple");
})();
